benchmark/src/main/java/com/linecorp/decaton/benchmark/RecordsGenerator.java

```diff
@@ -0,0 +1,92 @@
+
+import java.util.ArrayDeque;
+import java.util.Deque;
+import java.util.Properties;
+import java.util.Random;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Future;
+
+import org.apache.kafka.clients.producer.KafkaProducer;
+import org.apache.kafka.clients.producer.Producer;
+import org.apache.kafka.clients.producer.ProducerConfig;
+import org.apache.kafka.clients.producer.ProducerRecord;
+import org.apache.kafka.clients.producer.RecordMetadata;
+import org.apache.kafka.common.serialization.ByteArraySerializer;
+
+import lombok.extern.slf4j.Slf4j;
+
+@Slf4j
+public class RecordsGenerator {
+    public static final int LATENCY_CYCLE = 1000;
+
+    private final int[] latencies;
+
+    public RecordsGenerator(int maxProcessLatencyMs) {
+        latencies = new int[LATENCY_CYCLE];
+        // Use maxProcessingLatencyMs as the seed so that we always get consistent
+        // set of latencies.
+        Random random = new Random(maxProcessLatencyMs);
+        for (int i = 0; i < latencies.length; i++) {
+            latencies[i] = random.nextInt(maxProcessLatencyMs);
+        }
+    }
+
+    private static void consumeResults(Deque<Future<RecordMetadata>> results, boolean await)
+            throws InterruptedException {
+        Future<RecordMetadata> future;
+        while ((future = results.peekFirst()) != null) {
+            if (!future.isDone() && !await) {
+                break;
+            }
+            results.pollFirst();
+            try {
+                future.get();
+            } catch (ExecutionException e) {
+                throw new RuntimeException("failed producing record", e);
+            }
+        }
+    }
+
+    public void generate(String bootstrapServers, String topic, int numTasks) throws InterruptedException {
+        Properties props = new Properties();
+        props.setProperty(ProducerConfig.CLIENT_ID_CONFIG, "decaton-benchmark");
+        props.setProperty(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, bootstrapServers);
+        props.setProperty(ProducerConfig.ACKS_CONFIG, "all");
+        props.setProperty(ProducerConfig.LINGER_MS_CONFIG, "10");
+        props.setProperty(ProducerConfig.MAX_IN_FLIGHT_REQUESTS_PER_CONNECTION, "1");
+
+        log.info("Start generating {} tasks", numTasks);
+        Deque<Future<RecordMetadata>> results = new ArrayDeque<>();
+        try (Producer<byte[], Task> producer =
+                     new KafkaProducer<>(props, new ByteArraySerializer(), new Task.KafkaSerializer())) {
+            for (int i = 0; i < numTasks; i++) {
+                int latencyMs = latencies[i % latencies.length];
+                Task task = new Task(System.currentTimeMillis(), latencyMs);
+                ProducerRecord<byte[], Task> record = new ProducerRecord<>(topic, null, task);
+                Future<RecordMetadata> result = producer.send(record);
+                results.addLast(result);
+
+                consumeResults(results, false);
```

Comment:
Why consuming at each loop (though not awaiting) rather than "send and forget in loop => wait all after loop" ?

Reply:
Because we don't know how much will be specified for numTasks, taking that approach would increase memory occupation indefinitely I think.

Reply:
I see.